src/hotspot/share/gc/z/zInitialize.cpp

```diff
@@ -62,3 +73,51 @@ ZInitialize::ZInitialize(ZBarrierSet* barrier_set) {
 
   pd_initialize();
 }
+
+void ZInitialize::register_error(bool debug, const char *error) {
+  guarantee(!_finished, "Only register errors during initialization");
+
+  if (!_had_error) {
+    strncpy(_error_message, error, ErrorMessageLength - 1);
+    _had_error = true;
+  }
+
+  if (debug) {
+    log_error_pd(gc)("%s", error);
+  } else {
+    log_error_p(gc)("%s", error);
+  }
+}
+
+void ZInitialize::error(const char* msg_format, ...) {
+  va_list argp;
+  va_start(argp, msg_format);
+  const FormatBuffer<ErrorMessageLength> error(FormatBufferDummy(), msg_format, argp);
+  va_end(argp);
+  register_error(false /* debug */, error);
+}
+
+void ZInitialize::error_d(const char* msg_format, ...) {
+  va_list argp;
+  va_start(argp, msg_format);
+  const FormatBuffer<ErrorMessageLength> error(FormatBufferDummy(), msg_format, argp);
+  va_end(argp);
+  register_error(true /* debug */, error);
+}
+
+bool ZInitialize::had_error() {
+  return _had_error;
+}
+
+const char* ZInitialize::error_message() {
+  assert(had_error(), "Should have registered an error");
+  if (!had_error()) {
+    return "Unknown error, check error GC logs";
+  }
+  return _error_message;
```

Comment:
Maybe invert the conditional?
```suggestion
  if (had_error()) {
    return _error_message;
  }
  return "Unknown error, check error GC logs";
```